Objects/typeobject.c

```diff
@@ -343,6 +344,49 @@ _PyStaticType_GetBuiltins(void)
 
 /* end static builtin helpers */
 
+static void
+type_set_flags(PyTypeObject *tp, unsigned long flags)
+{
+#ifdef Py_GIL_DISABLED
+    if (tp->tp_flags & Py_TPFLAGS_READY) {
+        // Since PyType_HasFeature() reads the flags without holding the type
+        // lock, we need an atomic store here.
+        FT_ATOMIC_STORE_ULONG_RELAXED(tp->tp_flags, flags);
+    }
+    else {
+        // This is a small optimization to avoid the atomic stores in
+        // cases when new types are being initialized.  If type is not
+        // yet ready, a non-atomic store is okay.  When the type is
+        // finally marked ready then an atomic store will be done near
+        // the end of PyType_Read() (the READY flag is set and then the
+        // READYING flag is cleared after that).  That ensures any other
+        // thread reading the type flags will read the correct value.
+        tp->tp_flags = flags;
+    }
+#else
+    tp->tp_flags = flags;
+#endif
+}
+
+static void
+type_set_flags_with_mask(PyTypeObject *tp, unsigned long mask, unsigned long flags)
+{
+    ASSERT_TYPE_LOCK_HELD();
+    unsigned long new_flags = (tp->tp_flags & ~mask) | flags;
+    type_set_flags(tp, new_flags);
+}
+
+static void
+type_add_flags(PyTypeObject *tp, unsigned long flag)
+{
+    type_set_flags(tp, tp->tp_flags | flag);
```

Comment:
`type_add_flags` and `type_clear_flags` do not have an `ASSERT_TYPE_LOCK_HELD()`. Are they called without the type lock?

Reply:
Yes, they are sometimes called without the lock.  That should be done only in places were the type is being initially created and the lock will be held later, before the type is revealed.  At least, that's how I think it needs to work.

I would either try to fix that code so it takes the lock or I could make the assert something like `if <type ready> then ASSERT_TYPE_LOCK_HELD()`